---
title: "Random Forest: Original Dataset Cleaning Of Data"
author: "Matthew Curcio"
date: "August 31, 2017"
output: html_document
---

## Introduction - NOTES

For additional information see, [Building A Random Forest Classifier](https://www.biostars.org/p/86981/)  

[RPubs](http://rpubs.com/oaxacamatt/Random-Forest:Original-Dataset-Cleaning-Of-Data).

### Protein classes:

0. [Homo sapiens proteins - reviewed - NOT (1-6)]
1. [Erythrocruorin](https://en.wikipedia.org/wiki/Erythrocruorin)  
2. [Hemerythrin](https://en.wikipedia.org/wiki/Hemerythrin)  
3. [Hemocyanin](https://en.wikibooks.org/wiki/Structural_Biochemistry/Hemocyanin)  
4. [Leghemoglobin](https://en.wikipedia.org/wiki/Leghemoglobin)  
5. [Myoglobin](https://en.wikipedia.org/wiki/Myoglobin)  
6. [Hemoglobin](https://en.wikipedia.org/wiki/Hemoglobin)  

NOTE 1: Special attention was taken to reduce the set of Human proteins that were similar to the positive dataset by using Uniprot's advanced search feature.
[Human Proteins](http://www.uniprot.org/uniprot/?query=homo%20sapiens%20existence:%22evidence%20at%20protein%20level%22%20NOT%20myoglobin%20NOT%20erythrocruorin%20NOT%20hemoglobin%20NOT%20hemerythrin%20NOT%20hemocyanin%20NOT%20leghemoglobin&fil=&sort=score)
The difference between homo sapien reviewed proteins with and with out [1-6] is:
[146][mcc: uniprot_originals_existence_1]$ grep "^>" -c uniprot-taxonomy%3A-Homo+sapiens+%28Human%29+%5B9606%5D-+AND+reviewed%3Ayes+--.fasta 
20160  
[147][mcc: uniprot_originals_existence_1]$ grep "^>" -c uniprot-homo+AND+sapiens+AND+existence%3A-evidence+at+protein+level-+NOT+myo--.fasta 
16328
NOTE 2: The proteins used here only used proteins from Uniprot PE=1, (protein existence = evidence at protein level)


?? found in hemoglobin file line six: aa = 1156
sp|Q86VB7|C163A_HUMAN Scavenger receptor cysteine-rich type 1 protein M130 OS=Homo sapiens GN=CD163 PE=1 SV=2

https://www.r-bloggers.com/how-random-forests-improve-simple-regression-trees/

### Produce AA matrix:  

* Column 1 = Protein Class (0:1)  
  ** Protein files from Uniprot
  ** class = 0 indicates homo sapien protein reviewed NOT [1-6]  
  ** class = 1 indicates proteins [1-6]
* Column 2 = Total AA per protein  
* Columns 3:22; (G, P, A, V, L, I, M, C, F, Y, W, H, K, R, Q, N, E, D, S, T)  
* Columns 23:422; (GG, GP, GA, ... , TT)  


```{r libraires}
suppressMessages(library(seqinr))
suppressMessages(library(stringr))
suppressMessages(library(readr))
suppressMessages(library(dplyr))
```

### Read FASTA files, ONLY in .fasta, and produces .txt files.

```{r read_fasta, eval=TRUE}

fasta_2_sequences <- function() {
    UNIPROT_DIR = "~/Dropbox/git_projects/random_forest_project/cleaning_data_step_1"
    setwd(UNIPROT_DIR)
    protein_class = c("homo_sapiens", "erythrocruorin", "hemerythrin", 
                      "hemocyanin", "leghemoglobin", "myoglobin", "hemoglobin")
    o2_binders = list.files(pattern = ".fasta$")
    # o2_binders
    # length(o2_binders)
    
    # for loop to read sequences using seqinr:read.fasta
    for (i in seq(length(o2_binders))) {
    prot_seq <- read.fasta(file = o2_binders[i], seqtype = "AA", 
                           seqonly = TRUE, as.string = TRUE)
    file_name <- paste("protein_class_", protein_class[i], ".txt", sep = "")
    write.table(x = prot_seq, file = file_name, sep = "\n",
                col.names = FALSE, row.names = FALSE)
    }
}

# Run function
fasta_2_sequences()

```

### Function: percent_aa; takes a string of single letter amino acid abrevations and returns a vector of length (nrow = 1, ncol = 22) of %AA.
cell(1,1) is the protein class, (0:6),
** protein class(0) = null sett of human proteins, see above
cell(1,2) is total number of characters in string
cell(1,3:22) is percent AA

```{r monopeptides}

read_seq_txt <- function() {
    UNIPROT_DIR = "~/Dropbox/git_projects/random_forest_project/cleaning_data_step_1"
    setwd(UNIPROT_DIR)
    protein_class = c(1,2,3,6,0,4,5)
    o2_seqs = list.files(pattern = ".txt$")
    for (i in seq(length(o2_seqs))) {
        df = read.csv(file = o2_seqs[i], header = FALSE)
        df %>% mutate_if(is.factor, as.character) -> df
        df = t(df)
        percent_aa(df, protein_class[i])
    }
}

percent_aa <- function(df, protein_class) {
    col_titles = t(c("Class", "TotalAA", "G", "P", "A", "V", "L", "I", "M", "C",
                   "F", "Y", "W", "H", "K", "R", "Q", "N", "E", "D", "S", "T"))
    file_name <- paste("protein_class_", protein_class, "_aa.csv", sep = "")
    write.table(col_titles, file_name,  sep = ",", 
                col.names = FALSE, row.names = FALSE, eol = "\n")
    ############################################
    aa_nums = as.vector(matrix(0, ncol = 22))
    for (row in seq(length(df))) {
        # First column is protein_class
        aa_nums[1] = protein_class
        # Second column is total number of amino acids
        total_aa = nchar(df[1,row], keepNA = FALSE)
        aa_nums[2] = total_aa
        # Column 3:22 - Calculate percent AA
        aa_nums[3] = str_count(df[1,row], "G") / total_aa
        aa_nums[4] = str_count(df[1,row], "P") / total_aa
        aa_nums[5] = str_count(df[1,row], "A") / total_aa
        aa_nums[6] = str_count(df[1,row], "V") / total_aa
        aa_nums[7] = str_count(df[1,row], "L") / total_aa
        aa_nums[8] = str_count(df[1,row], "I") / total_aa
        aa_nums[9] = str_count(df[1,row], "M") / total_aa
        aa_nums[10] = str_count(df[1,row], "C") / total_aa
        aa_nums[11] = str_count(df[1,row], "F") / total_aa
        aa_nums[12] = str_count(df[1,row], "Y") / total_aa
        aa_nums[13] = str_count(df[1,row], "W") / total_aa
        aa_nums[14] = str_count(df[1,row], "H") / total_aa
        aa_nums[15] = str_count(df[1,row], "K") / total_aa
        aa_nums[16] = str_count(df[1,row], "R") / total_aa
        aa_nums[17] = str_count(df[1,row], "Q") / total_aa
        aa_nums[18] = str_count(df[1,row], "N") / total_aa
        aa_nums[19] = str_count(df[1,row], "E") / total_aa
        aa_nums[20] = str_count(df[1,row], "D") / total_aa
        aa_nums[21] = str_count(df[1,row], "S") / total_aa
        aa_nums[22] = str_count(df[1,row], "T") / total_aa
        aa_nums = t(aa_nums)
        # Write/append vector of AA values
        write(aa_nums, file = file_name, append = TRUE, ncolumns = 22, sep = ",")
    }
}

# Generate single AA percent compositions
read_seq_txt()

```

### At this point, the MONO-peptide files have been produced.

## Continue on to DI-peptide production using file, [name.Rmd] at (location)
