---
title: "Transforming Protein Dataset(s) to %AA"
author: "Matthew Curcio"
date: "Feb 26, 2019"
output: html_document
---

  


### Produce AA matrix:  

* Column 1 = Protein Class (0:1)  
  ** Protein files from Uniprot
  ** class = 0 indicates homo sapien protein reviewed NOT [1-6]  
  ** class = 1 indicates proteins [1-6]
* Column 2 = Total AA per protein  
* Columns 3:22; (G, P, A, V, L, I, M, C, F, Y, W, H, K, R, Q, N, E, D, S, T)  


```{r libraires}
suppressMessages(library(seqinr))
suppressMessages(library(stringr))
suppressMessages(library(readr))
suppressMessages(library(dplyr))
```

### Read FASTA files, ONLY in .fasta, and produces .txt files.

```{r read_fasta, eval=TRUE}

fasta_2_sequences <- function() {
    UNIPROT_DIR = "/FASTA_files"
    setwd(UNIPROT_DIR)
    protein_class = c("homo_sapiens", "erythrocruorin", "hemerythrin", 
                      "hemocyanin", "leghemoglobin", "myoglobin", "hemoglobin")
    o2_binders = list.files(pattern = ".fasta$")
    # o2_binders
    # length(o2_binders)
    
    # for loop to read sequences using seqinr:read.fasta
    for (i in seq(length(o2_binders))) {
    prot_seq <- read.fasta(file = o2_binders[i], seqtype = "AA", 
                           seqonly = TRUE, as.string = TRUE)
    file_name <- paste("protein_class_", protein_class[i], ".txt", sep = "")
    write.table(x = prot_seq, file = file_name, sep = "\n",
                col.names = FALSE, row.names = FALSE)
    }
}

# Run function
fasta_2_sequences()

```

### Function: percent_aa; takes a string of single letter amino acid abrevations and returns a vector of length (nrow = 1, ncol = 22) of %AA.
cell(1,1) is the protein class, (0:6),
** protein class(0) = null sett of human proteins, see above
cell(1,2) is total number of characters in string
cell(1,3:22) is percent AA

```{r producePercentAA}

# Reads .txt file and converts it to dataframe then call function 'percent_aa'
read_seq_txt <- function() {
    UNIPROT_DIR = "/home/mcc/Dropbox/ml-o2-binding-proteins/2-pipeline/Feb2019_protein_classes/single_line_aa_seq"
    setwd(UNIPROT_DIR)
    protein_class = c("Ctrl", "Ery", "Hcy", "Hhe", "Hgb", "Lgb", "Mgb") 
    o2_binders = list.files(pattern = ".txt$")
    for (i in seq(length(o2_binders))) {
        df = read.csv(file = o2_binders[i], header = FALSE)
        df %>% mutate_if(is.factor, as.character) -> df
        df = t(df)
        percent_aa(df, protein_class[i])
    }
}

# Takes 
percent_aa <- function(df, protein_class) {
    title = t(c("Class", "TotalAA", "G", "P", "A", "V", "L", "I", "M", "C",
                     "F", "Y", "W", "H", "K", "R", "Q", "N", "E", "D", "S", "T"))
    file_name <- paste("protein_class_", protein_class, "_aa.csv", sep = "")
    write.table(title, file_name, sep = ",", col.names = FALSE, 
                row.names = FALSE, eol = "\n")
    ############################################
    aa_nums = as.vector(matrix(0, ncol = 22))
    for (row in seq(length(df))) {
        # First column is protein_class
        aa_nums[1] = protein_class[1]
        # Second column is total number of amino acids
        total_aa = nchar(df[1,row], keepNA = FALSE)
        aa_nums[2] = total_aa
        # Column 3:22 - Calculate percent AA
        for (i in c(3:22)) {
            aa_nums[i] = str_count(df[1,row], title[i]) / total_aa
        }
        aa_nums = t(aa_nums)
        # Write/APPEND vector of AA values
        write(aa_nums, file = file_name, append = TRUE, ncolumns = 22, sep = ",")
    }
}

# Generate single AA percent compositions
read_seq_txt()

```

### At this point, the MONO-peptide files have been produced.

## Continue on to DI-peptide production using file, [name.Rmd] at (location)
