---
title: "2-18-19-aa-2-percent-comp.rmd"
author: "Matthew Curcio"
date: "February 18, 2019"
output: html_document
---

## Title: 2-18-19-aa-2-percent-comp.rmd

#### Purpose: Produce AA matrix:  

1. This script takes as input seven .txt files marked as "protein_class_$var.txt".
The varialbe name is one of the seven classes listed below; 

- Protein Class = ("Ctrl", "Ery", "Hcy", "Hhe", "Hgb", "Lgb", "Mgb")

2. The first function determines which lines are less than 50 amino acids in length
and deletes these lines.

3. The script then reads the files line by line and computes the percent amino acid composition of the sequence string. 

4. The output of this script is a .csv file, which is written line by line as it calculates.  The rows represent a polypeptide. Each row in the .csv file are setup in 22 columns, listed below;

- Column 1; Protein Class = ("Ctrl", "Ery", "Hcy", "Hhe", "Hgb", "Lgb", "Mgb") 
- Column 2; Total AA per protein  
- Columns 3:22; aa = ("G", "P", "A", "V", "L", "I", "M", "C", "F", "Y", 
                      "W", "H", "K", "R", "Q", "N", "E", "D", "S", "T")  

5. The 7 .csv files will be concatenated using the bash function `cat`. 

This completes the transformation of the polypeptide sequences into percent composition of the 20 amino acids for the seven grouped proteins.

```{r libraries include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

Libraries = c("readr", "seqinr", "stringr", "dplyr", "beepr")

# Install if not present
for(p in Libraries){
    if(!require(p, character.only = TRUE))
        install.packages(p)
    library(p, character.only = TRUE)
}
```

## Import Data



### Read FASTA files, ONLY in .fasta, and produces .txt files.

```{r read_fasta, eval=FALSE, include=FALSE}

fasta_2_sequences <- function() {
    UNIPROT_DIR = "/home/mcc/Desktop/ml-o2-binding-proteins/2-pipeline/Feb2019_protein_classes"
    setwd(UNIPROT_DIR)
    protein_class = c("Ctrl", "Ery", "Hcy", "Hhe", "Hgb", "Lgb", "Mgb") 
    o2_binders = list.files(pattern = ".txt$")
    
    # for loop to read sequences using seqinr:read.fasta
    for (i in seq(length(o2_binders))) {
    prot_seq <- read.fasta(file = o2_binders[i], 
                           seqtype = "AA", 
                           seqonly = TRUE, as.string = TRUE)
    file_name <- paste("protein_class_", protein_class[i], ".txt", sep = "")
    write.table(x = prot_seq, file = file_name, sep = "\n",
                col.names = FALSE, row.names = FALSE)
    }
}

# Run function
fasta_2_sequences()

```

### Function: monopeptides - takes .txt files and outputs amino acid percent composition into a .csv file format.

- Column 1; Protein Class = ("Ctrl", "Ery", "Hcy", "Hhe", "Hgb", "Lgb", "Mgb") 
- Column 2; Total AA per protein  
- Columns 3:22; aa = ("G", "P", "A", "V", "L", "I", "M", "C", "F", "Y", 
                      "W", "H", "K", "R", "Q", "N", "E", "D", "S", "T")  

```{r monopeptides}
read_seq_txt <- function() {
    UNIPROT_DIR = "/home/mcc/Desktop/ml-o2-binding-proteins/2-pipeline/Feb2019_protein_classes"
    setwd(UNIPROT_DIR)
    protein_class = c("Ctrl", "Ery", "Hcy", "Hhe", "Hgb", "Lgb", "Mgb") 
    o2_seqs = list.files(pattern = ".txt$")
    for (i in seq(length(o2_seqs))) {
        df = read.csv(file = o2_seqs[i], header = FALSE)
        df %>% mutate_if(is.factor, as.character) -> df
        df = t(df)
        percent_aa(df, protein_class[i])
    }
}
```


```{r percent_aa}
percent_aa <- function(df, protein_class) {
    col_titles = t(c("Class", "TotalAA", "G", "P", "A", "V",
                     "L", "I", "M", "C", "F", "Y", "W", "H", 
                     "K", "R", "Q", "N", "E", "D", "S", "T"))
    file_name <- paste("pro_class_", protein_class, "_aa.csv", sep = "")
    write.table(col_titles, 
                file_name,  
                sep = ",", 
                col.names = FALSE, 
                row.names = FALSE, 
                eol = "\n")
    ############################################
    aa_nums = as.vector(matrix(0, ncol = 22))
    for (row in seq(length(df))) {
        # First column is protein_class
        aa_nums[1] = protein_class
        # Second column is total number of amino acids
        total_aa = nchar(df[1,row], keepNA = FALSE)
        aa_nums[2] = total_aa
        # Column 3:22 - Calculate percent AA
        aa_nums[3] = str_count(df[1,row], "G") / total_aa
        aa_nums[4] = str_count(df[1,row], "P") / total_aa
        aa_nums[5] = str_count(df[1,row], "A") / total_aa
        aa_nums[6] = str_count(df[1,row], "V") / total_aa
        aa_nums[7] = str_count(df[1,row], "L") / total_aa
        aa_nums[8] = str_count(df[1,row], "I") / total_aa
        aa_nums[9] = str_count(df[1,row], "M") / total_aa
        aa_nums[10] = str_count(df[1,row], "C") / total_aa
        aa_nums[11] = str_count(df[1,row], "F") / total_aa
        aa_nums[12] = str_count(df[1,row], "Y") / total_aa
        aa_nums[13] = str_count(df[1,row], "W") / total_aa
        aa_nums[14] = str_count(df[1,row], "H") / total_aa
        aa_nums[15] = str_count(df[1,row], "K") / total_aa
        aa_nums[16] = str_count(df[1,row], "R") / total_aa
        aa_nums[17] = str_count(df[1,row], "Q") / total_aa
        aa_nums[18] = str_count(df[1,row], "N") / total_aa
        aa_nums[19] = str_count(df[1,row], "E") / total_aa
        aa_nums[20] = str_count(df[1,row], "D") / total_aa
        aa_nums[21] = str_count(df[1,row], "S") / total_aa
        aa_nums[22] = str_count(df[1,row], "T") / total_aa
        aa_nums = t(aa_nums)
        # Write/append vector of AA values
        write(aa_nums, file = file_name, append = TRUE, ncolumns = 22, sep = ",")
    }
}
```

```{r}
# Generate single AA percent compositions
read_seq_txt()
beep(sound = 1, expr = NULL)
```

### At this point, the MONO-peptide files have been produced.

## Continue to Exploratory Data Analysis.

